todo_app/app.py

```diff
@@ -1,38 +1,82 @@
-from pprint import pprint
+import os
+import requests
 from flask import Flask, render_template, request, redirect
-from todo_app.data.session_items import create_card, get_lists, get_cards, move_card, remove_card, ViewModel
+from datetime import datetime
+from todo_app.data.session_items import check_default_lists, create_card, get_lists, get_cards, move_card, remove_card, ViewModel
+from todo_app.user import TestUser,User,Anonymous
+from flask_login import LoginManager, login_required, UserMixin, login_user, current_user
+from oauthlib.oauth2 import WebApplicationClient
 
-from todo_app.flask_config import Config
+login_manager = LoginManager()
+
+@login_manager.unauthorized_handler
+def unauthenticated():
+    github_client =  WebApplicationClient(os.environ.get('OAUTH_CLIENT_ID'))
+    github_redirect = github_client.prepare_request_uri("https://github.com/login/oauth/authorize")
+
+    return redirect(github_redirect)
+
+@login_manager.user_loader
+def load_user(user_id):
+    return User(user_id)
 
 def create_app():
     app = Flask(__name__)
-    app.config.from_object(Config)
+    app.secret_key = os.getenv('OAUTH_CLIENT_SECRET')
+    app.config['LOGIN_DISABLED'] = True
+    login_manager.init_app(app)
+    if os.getenv('FLASK_ENV') == 'e2e_Test':
+        login_manager.anonymous_user = TestUser
+    else:
+        login_manager.anonymous_user = Anonymous
+
+    check_default_lists()
 
     @app.route('/')
+    @login_required
     def index():
         item_view_model = ViewModel(get_cards(),get_lists())
-        return render_template('index.html', view_model=item_view_model)
+        return render_template('index.html',view_model=item_view_model,user_role=current_user.role)
 
 
     @app.route('/',methods=['POST'])
+    @login_required
     def test():
-        name=request.form.get("card_name")
-        create_card(name,request.form['submit_button'])
-        return redirect('/')
+        if 'LOGIN_DISABLED' in app.config or current_user.role == 'WRITER':
+            create_card(request.form.get("card_desc"),request.form.get('submit_button'),request.form.get('card_name'),request.form.get('card_due'))
+            return redirect('/')
```

Comment:
Currently when the user is unauthorised, you aren't returning anything from this function. That would result in a 500 (internal server error).

Would be nicer to return a suitable error code like 401 or 403 with a useful message.

Reply:
And instead of repeating the if/else logic on each route, you could try writing a decorator function.